### ISSUE TITLE
Add option for HTTP Authentication

### ISSUE THREAD

**Issue by fletchowns** [memen45/SubMusic#46]
I use HTTP Authentication for my subsonic installation rather than the standard subsonic username & password login, so that I can easily utilize fail2ban to throttle & block invalid login attempts.

I added SubMusic to my Venu 2 Plus, specified my server URL, username, and password, but I'm getting an error when I click Test Server:
```
GarminSdkError::-400
Error HTTP -->
GarminSdkError::"
INVALID_HTTP_BODY_IN_NETWORK_RESPONSE"
```

Subsonic itself supports HTTP Authentication, it would be great if SubMusic did too!

I am only theorizing that HTTP Authentication is the culprit here, I'm not familiar enough with Garmin Connect IQ Apps to debug further. I'd be happy to sponsor implementation of this feature though, if that is indeed the cause of the issue.

**Comment by memen45** [comment]
Hi, thanks for your comment! I checked the [Subsonic API spec](http://www.subsonic.org/pages/api.jsp), but I could not find any mention of HTTP authentication there. Do you apply the http authentication on top of the authentication scheme as in the Subsonic API specification?

**Comment by fletchowns** [comment]
Thanks for the reply! Interesting that it is not mentioned in the spec. It seems to be a feature that Subsonic supports out of the box, it has been a little while since I set this up though so maybe I am not remembering some key detail. I don't think there was any special configuration I made in Subsonic to enable it as far as I can tell.

I use nginx to implement HTTP Authentication for my subsonic installation (trimmed this snippet down for clarity):
```
server {
    ...
    location / {
        auth_basic "Restricted";
        auth_basic_user_file /path/to/.subsonic_htpasswd_users;

        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
        proxy_set_header X-Forwarded-Host $http_post;
        proxy_set_header Host $http_host;
        proxy_max_temp_file_size 0;

        proxy_pass http://localhost:58383/;
        proxy_redirect http:// https://;
    }
}
```

Both `/path/to/.subsonic_htpasswd_users` and subsonic itself have users defined with identical credentials. When I navigate to my subsonic URL in a browser, the browser prompts me for HTTP authentication credentials, and then it logs me right in, skipping the normal Subsonic login page entirely.

[dSub](https://play.google.com/store/apps/details?id=github.daneren2005.dsub), the android app I use with my subsonic server, has an option for toggling whether the HTTP authentication headers are included or not:
![image](https://user-images.githubusercontent.com/153107/172459227-c7c4d6e9-5267-4848-9b4e-7dff70563a78.png)

There's some discussion of adding that option over in here: https://github.com/daneren2005/Subsonic/issues/870

**Comment by memen45** [comment]
Is the http authentication as simple as adding a header to each request?
```
Authorization: Basic <base64(username)>:<base64(password)>
```
Given that the Subsonic API requires the username/password for each request, and does not have a session key, this header would be required for each API request right?

Then I think it should be possible to add a SubMusic setting similar to DSub, to include this header in the request if enabled. This would only take affect in the Subsonic API backend, as I am not aware of a similar feature for Ampache.

**Comment by fletchowns** [comment]
Yup it does look like it might be that simple! Though I think it may be more along the lines of:
```
Authorization: Basic <base64(username:password)>
```

Though my test of calculating this with a CLI command and comparing it to what my browser is actually sending in the Network tab I'm still off by one character (an extra equals sign at the end of the value...) and I'm not sure why yet.

**Comment by Ramrawd** [comment]
I just wanted to chime in and say I'm also getting this same exact error message. 

I'm using an instance of NextCloud that's piped through nginx as well. 

Thanks!

**Comment by memen45** [comment]
Due to holidays coming week it will take a little longer to implement and test this feature and issue the update. Sorry and thanks for waiting!

Notes:
- [ ] add setting 'Enable HTTP basic authentication', default off
- [ ] add default header to Subsonic API class and class functions (empty header)
- [ ] add conditional initialisation to constructor
- [ ] testing

**Comment by memen45** [comment]
Implemented in 047b22240f7b2f55be3f61b68f7c2c09f0f56883 and part of the update [0.1.25-zulu](https://github.com/memen45/SubMusic/tree/0.1.25-zulu).

I am not able to test the feature, as I do not have HTTP Basic authentication enabled on my server. Please report any issues you may have!

**Comment by fletchowns** [comment]
Very exciting!! Thanks so much for adding this new functionality.

I tried out the new version and after enabling the new setting, I'm able to see the list of my playlists under More -> Browser -> Playlists. However, when I toggle one of them to be available offline, and then run the sync, my watch displays `HttpError::401`. Additionally, if I click Test Server, my watch says `HttpError::401 Error HTTP --> HttpError::Unauthorized`.

So it at least appears to be partially working, since I was able to see my list of playlists from my watch, but not fully working as I wasn't able to sync any songs.

Thanks again for your efforts with this! I was happy to make a donation to show my appreciation of the project and support your efforts.

**Comment by memen45** [comment]
Thanks a lot for your generosity and appreciation! 

Some methods were missing the authentication header, it should be fixed now with the 0.2.0 update! It could take a couple of hours for the update to become available.

**Comment by fletchowns** [comment]
Wow thanks for the quick turnaround! I am thrilled to report that everything is working great for me with the new update. Test Server, Sync, Browse, it's all looking good now. Nice work on this feature! Great work on the app overall as well, I am very glad you created it!
